### PR TITLE
Make a stand-in for `std::expected` available in the code base

### PR DESCRIPTION
**Pull request by scottschurr**
## Make a Stand-in For `std::expected` Available in the Code Base

`std::expected` is a proposed type for the standard library.  The current paper for `std::expected` is: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0323r10.html

The interface of `std::expected` looks great to me.  It's like a `std::optional` that tells you _why_ something came back empty.  I'd like to be able to use it in the code base without waiting for us to adopt C++23.

### Context of Change

Boost does not directly provide an interface like `std::expected`.  However Boost has something similar called `boost::outcome_v2::result`.  I've coded a fairly thin wrapper around `boost::outcome_v2::result` that provides the characteristics of `std::expected` that I particularly like.  I've named that wrapper `ripple::Expected`.

That wrapper, and unit tests for it, are in the first commit.

My initial impression was that `Expected` would be immediately usable in significant portions of the code base.  So I integrated `Expected` into our code base.  To my surprise I found relatively few places where I could directly drop in `Expected` as a replacement for returning a `std::pair` where one of the pair members was the reason for the failure.  But I did find a few.  You can find that integration in the second commit.

Reviewers should consider two things:
- Given that `Expected` is not essential to the code base, should we still incorporate it?  I think we should.  But you may not agree
- If it is worth incorporating `Expected` into the code base, do you see any coding errors or places for improvement?

For what it's worth, the `std::expected` paper is on track to go into C++23 and has passed the Library Evolution working group.  But there's some risk because it has not yet survived the Library (wording) working group.

### Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)

This is an internals-only change.  It does not need to be mentioned in the release notes.

## Before / After

From an interface perspective, an implementation returning an `Expected` changes something sorta like this ...
```
std::pair<std::optional<AcctRootRd>, NotTEC >
makeAcctRoot(std::shared_ptr<STLedgerEntry const> slePtr)
{
    if (NotTEC const ter = validateAcctRootSle(slePtr); !isTesSuccess(ter))
        return {nullopt, ter}

    return {AcctRootRd(std::move(slePtr), tesSUCCESS};
}
```
... to this ...
```
Expected<AcctRootRd, NotTEC>
makeAcctRoot(std::shared_ptr<STLedgerEntry const> slePtr)
{
    if (NotTEC const ter = validateAcctRootSle(slePtr); !isTesSuccess(ter))
        return Unexpected(ter);

    return AcctRootRd(std::move(slePtr));
}
```
At the call site this ...
```
    auto const acctRoot = makeAcctRoot(ctx.view.read(keylet::account(id)));
    if (!isTesSuccess(acctRoot.second))
        return acctRoot.second;

    std::uint32_t const uFlagsIn = {acctRoot.first->flags()};
```
... becomes this ...
```
    auto const acctRoot = makeAcctRoot(ctx.view.read(keylet::account(id)));
    if (!acctRoot)
        return acctRoot.error();

    std::uint32_t const uFlagsIn = {acctRoot->flags()};
```
The examples have not been run through a compiler, so there may be bugs.

## Test Plan

Unit tests for `Expected` are included.  The integration also builds confidence that the `Expected` wrapper is behaving as expected.

## Future Tasks

Presuming `std::expected` makes it into C++23 it will make sense to remove `ripple::Expected` from the code base and replace its usage with `std::expected`.  There _shouldn't_ be significant usage changes to `std::expected`, since it has passed through Library Evolution.  So the replacement should be straight forward at that time.

### CODE REVIEW

**Comment by scottschurr** [comment]
I pushed a commit with a partial specialization for `Expected<void, E>`.  To me it looks like it should serve the purpose.  Let me know what you think.  Thanks.

File: src/ripple/basics/Expected.h
Comment (seelabs) [inline src/ripple/basics/Expected.h:215]: It looks like the proposed standard `expected` has a default constructor for a `void` return (ref: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0323r9.html#expected.object.ctor). I think a default constructor makes sense in this case, and give the standard has it, I'd like to remove this constructor and replace it with a default constructor.

**Review by seelabs — COMMENTED** [review]
Thanks for coding up the `void` specialization. I had one comment about removing the `true_type` constructor and replacing it with a default constructor. Other than that it looks great.

File: src/ripple/basics/Expected.h
Comment (scottschurr) [reply src/ripple/basics/Expected.h]: Thanks for catching the default constructor for `Expected<void, E>`.  I looked for something like `boost::outcome_v2::success()` in the paper and didn't find it, so I punted.  Matching the currently proposed version of the paper seems like the best approach.  I've included an additional `[FOLD]` to make this change.

File: src/ripple/basics/base_uint.h
Comment (seelabs) [inline src/ripple/basics/base_uint.h:252]: Why does `badChar` assert but `badLength` doesn't assert? Consider losing the `assert`. It's changing the behavior of debug builds, and the change doesn't involve the major focus of this PR - which is `expected`. I think it's better if PRs focus on one thing only.

Having said that, I only weakly object. If you feel strongly, then it's your call.

**Review by seelabs — APPROVED** [review]
:+1: Great job on this!

I'm approving this now, but please see my comment about adding the `assert`. I'm approving now because however you want to resolve that is fine with me (including keeping it as-is).

File: src/ripple/basics/base_uint.h
Comment (scottschurr) [reply src/ripple/basics/base_uint.h:252]: The assert was intended to be a way to guarantee that all of the non-`okay` `ParseResult` values are covered.  We can see that by inspection now.  But in case another error code was added in the future the `assert` would let us know that the `Throw` had the wrong text.  But I don't feel strongly about it.  I'm fine with removing the `assert`.

**Review by seelabs — APPROVED** [review]
Still :+1: from me

**Review by thejohnfreeman — APPROVED** [review]

**Comment by scottschurr** [comment]
I squashed and pushed and expected CI to run.  Apparently Travis needs a payment before CI will run: https://app.travis-ci.com/github/ripple/rippled/requests

I'll wait until CI has passed before marking the pull request as passed.

**Comment by scottschurr** [comment]
Re-pushed to kick Travis into gear.

**Pull request closed without merging**
